docs/rfcs/040-Endpoint-Persistent-Unlogged-Files-Storage.md

```diff
@@ -0,0 +1,177 @@
+```mermaid
+sequenceDiagram
+    autonumber
+    participant co as Compute
+    participant ep as EPUFS
+    participant s3 as S3
+
+    co->>+ep: Store Unlogged Persistent File
+    opt is authenticated
+        ep->>s3: Write UPF to S3
+    end
+    ep->>-co: OK / Failure / Auth Failure
+```
+
+Read data:
+```mermaid
+sequenceDiagram
+    autonumber
+    participant co as Compute
+    participant ep as EPUFS
+    participant s3 as Storage
+
+    co->>+ep: Read Unlogged Persistent File
+    opt is authenticated
+        ep->>+s3: Request UPF from storage
+        s3->>-ep: Receive UPF from storage
+    end
+    ep->>-co: OK(response) / Failure(storage, auth, ...)
+```
+
+Compute Startup:
+```mermaid
+sequenceDiagram
+    autonumber
+    participant co as Compute
+    participant ps as Pageserver
+    participant ep as EPUFS
+    participant es as Extension server
+
+    note over co: Bind endpoint ep-xxx
+    par Get basebackup
+        co->>+ps: Request basebackup @ LSN
+        ps-)ps: Construct basebackup
+        ps->>-co: Receive basebackup TAR @ LSN
+    and Get startup-critical Unlogged Persistent Files
+        co->>+ep: Get all UPFs of endpoint ep-xxx
+        ep-)ep: Retrieve and gather all UPFs
+        ep->>-co: TAR of UPFs
+    and Get startup-critical extensions
+        loop For every startup-critical extension
+            co->>es: Get critical extension
+            es->>co: Receive critical extension
+        end
+    end
+    note over co: Start compute
+```
+
+CPlane ops:
+```mermaid
+sequenceDiagram
+    autonumber
+    participant cp as Control Plane
+    participant ep as EPUFS
+    participant s3 as Storage
+
+    alt Tenant deleted
+        cp-)ep: Tenant deleted
+        loop
+            ep->>s3: Remove data of deleted tenant from Storage
+        end
+        opt
+            ep-)cp: Tenant cleanup complete
+        end
+    else Endpoint reassigned
+        cp->>+ep: Endpoint reassigned
+        loop For every object associated with reassigned endpoint
+            ep->>s3: Remove data from Storage
+        end
+        ep->>-cp: Cleanup complete
+    else Endpoint removed
+        cp->>+ep: Endpoint removed
+        loop For every object associated with removed endpoint
+            ep->>s3: Remove object from Storage
+        end
+        ep->>-cp: Cleanup complete
+    end
+```
+
+### Scalability (if relevant)
+TBD
+
+Provisionally:  As this service is going to be part of compute startup, this
+service should be able to quickly respond to all requests.  Local caching of
+frequently restarted endpoints' data may be needed for best performance.
+
+### Security implications (if relevant)
+This service must be able to authenticate users at least by Tenant ID,
+preferably also Timeline ID and/or Endpoint ID.
+
+There may be a use case for transferring data from one endpoint to another,
+but that's currently not in scope.
+
+### Unresolved questions (if relevant)
+TBD
+
+## Alternative implementation (if relevant)
```

Comment:
The main alternative I'd be curious about is having computes talk to S3 directly instead of going through a separate service

Reply:
I think in this RFC it is somewhat implied that it is overly tricky on the auth side. But worth spelling that out:
- what are downsodes of bucket-per-tenant? cost, ops, etc
- can we have path per tenant without proxy service?

Reply:
> what are the downsides of bucket-per-tenant? cost, ops, etc

Also curious about this part. My gut feeling is that we will hit some AWS acc limits or provisioning issues if we will decide to provision hundreds of thousands of buckets. Also, the latency is the question here -- right now, we create a project, and compute/Postgres is ready to go within a couple of seconds. I doubt that provisioning an S3 bucket is fast, so that means it should be non-blocking and compute will run for an arbitrary amount of time without access to S3

> can we have path per tenant without proxy service?

Yeah, I thought about it too. I guess it's possible with a combination of IAM and k8s service account, like https://repost.aws/de/knowledge-center/s3-folder-user-access. Yet, as I got from Em, our NeonVMs do not work with service accounts currently, and it's tricky to implement. Either way, latency could be an issue again, i.e. provision/create IAM and SA. It will likely require some non-trivial cplane and compute work to make it non-blocking

Everything is worth mentioning, though, +1

Reply:
+1 for the proxy approach. It gives us more control over what happens. The proxy service can be pretty simple, with no local state so that it can be scaled and made highly available easily.

The API for the proxy can perhaps be an S3-compatible PUT/GET API, so that you can use standard clients libraries with it. Not sure how important that is, or if that's any better than a plain REST API.